src/datastore.jl

```diff
@@ -19,6 +40,58 @@ using .Threads
 const datastore = Dict{Int,RefState}()
 const datastore_lock = Ref{Union{Nothing, ReentrantLock}}(nothing)
 const id_counter = Ref{Union{Nothing, Atomic{Int}}}(nothing)
+const datastore_counter = Dict{Tuple{Int,Int}, Atomic{Int}}()
+const local_datastore_counter = Dict{Tuple{Int,Int}, Atomic{Int}}()
+
+function poolref(d::DRef)
+    with_datastore_lock() do
+        ctr = get!(local_datastore_counter, (d.owner,d.id)) do
+            # We've never seen this DRef, so tell the owner
+            if d.owner == myid()
+                poolref_owner(d)
+            else
+                @async remotecall(poolref_owner, d.owner, d)
```

Comment:
Do we need to lock the local datastore to do this?

Reply:
Technically we only need the read-write logic of `get!` to be locked, since `Dict` isn't thread-safe. If you know of a thread-safe implementation of `Dict`, we could use that to completely avoid locking the datastore for all of `poolref` and most of `poolunref`.